beater/agent_config_handler.go

```diff
@@ -0,0 +1,106 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package beater
+
+import (
+	"net/http"
+
+	"github.com/elastic/apm-server/convert"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/kibana"
+
+	"github.com/elastic/apm-server/agentcfg"
+)
+
+func agentConfigHandler(client *kibana.Client, secretToken string) http.Handler {
+
+	var handler http.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		send := wrap(w, r)
+		clientEtag := r.Header.Get("If-None-Match")
+
+		query, requestErr := buildQuery(r)
+		cfg, upstreamEtag, internalErr := agentcfg.Fetch(client, query, requestErr)
+
+		switch {
+
+		case requestErr != nil:
+			send(requestErr.Error(), http.StatusBadRequest)
+
+		case query == agentcfg.Query{}:
+			send(nil, http.StatusMethodNotAllowed)
+
+		case internalErr != nil:
+			send(internalErr.Error(), http.StatusInternalServerError)
+
+		case len(cfg) == 0:
+			send(nil, http.StatusNotFound)
+
+		case clientEtag != "" && clientEtag == upstreamEtag:
+			w.Header().Set("Cache-Control", "max-age=0")
+			send(nil, http.StatusNotModified)
+
+		case upstreamEtag != "":
+			w.Header().Set("Cache-Control", "max-age=0")
+			w.Header().Set("Etag", upstreamEtag)
+			fallthrough
+
+		default:
+			send(cfg, http.StatusOK)
+		}
```

Comment:
nit: would you mind removing the empty lines in the switch/case statement? I find that really hard to read.

Reply:
really spacing hurts your readability more than all lines crammed together without any breathing? 
hmm, ok. i am totally the opposite, then

Reply:
I like to use spaces to group things together, in the switch/case block I find it harder to read. But as I said that's a nit, if you prefer otherwise it is fine with me.